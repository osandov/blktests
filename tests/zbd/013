#!/bin/bash
# SPDX-License-Identifier: GPL-3.0+
# Copyright (C) 2025 Google LLC

. tests/zbd/rc
. common/null_blk

DESCRIPTION="test stacked drivers and queue freezing"
QUICK=1

requires() {
	_have_driver dm-crypt
	_have_fio
	_have_module null_blk
	_have_program cryptsetup
}

# Trigger blk_mq_freeze_queue() repeatedly because there is a bug in the
# Linux kernel 6.10..6.14 zoned block device code that triggers a deadlock
# between zoned writes and queue freezing.
queue_freeze_loop() {
	while true; do
		echo 4 >"$1"
		sleep .1
		echo 8 >"$1"
		sleep .1
	done
}

test() {
	set -e

	echo "Running ${TEST_NAME}"

	_init_null_blk nr_devices=0 queue_mode=2

	# A small conventional block device for the LUKS header.
	local null_blk_params=(
		blocksize=4096
		completion_nsec=0
		memory_backed=1
		size=4            # MiB
		submit_queues=1
		power=1
	)
	_configure_null_blk nullb0 "${null_blk_params[@]}"
	local hdev=/dev/nullb0

	# A larger zoned block device for the data.
	local null_blk_params=(
		blocksize=4096
		completion_nsec=0
		memory_backed=1
		size=1024         # MiB
		submit_queues=1
		zoned=1
		power=1
	)
	_configure_null_blk nullb1 "${null_blk_params[@]}"
	local zdev=/dev/nullb1

	local luks_passphrase=this-passphrase-is-not-secret
	{ echo "${luks_passphrase}" |
		  cryptsetup luksFormat --batch-mode ${zdev} \
			     --header ${hdev}; }
	local luks_vol_name=zbd-013
	{ echo "${luks_passphrase}" |
		  cryptsetup luksOpen \
			     --batch-mode "${zdev}" "${luks_vol_name}" \
			     --header ${hdev}; }
	local luksdev="/dev/mapper/${luks_vol_name}"
	local dmdev
	dmdev="$(basename "$(readlink "${luksdev}")")"
	ls -ld "${hdev}" "${zdev}" "${luksdev}" "/dev/${dmdev}" >>"${FULL}"
	local zdev_basename
	zdev_basename=$(basename "$zdev")
	local max_sectors_zdev
	max_sectors_zdev=/sys/block/"${zdev_basename}"/queue/max_sectors_kb
	echo 4 > "${max_sectors_zdev}"
	echo "${zdev_basename}: max_sectors_kb=$(<"${max_sectors_zdev}")" >>"${FULL}"
	local max_sectors_dm
	max_sectors_dm=/sys/block/"${dmdev}"/queue/max_sectors_kb
	echo "${dmdev}: max_sectors_kb=$(<"${max_sectors_dm}")" >>"${FULL}"
	queue_freeze_loop /sys/block/"$dmdev"/queue/read_ahead_kb &
	local loop_pid=$!
	local fio_args=(
		--bs=64M
		--direct=1
		--filename="${luksdev}"
		--runtime="$TIMEOUT"
		--time_based
		--zonemode=zbd
	)
	if ! _run_fio_verify_io "${fio_args[@]}" >>"${FULL}" 2>&1; then
		fail=true
	fi

	set +e

	kill "${loop_pid}"
	cryptsetup luksClose "${luks_vol_name}"
	_exit_null_blk

	if [ -z "$fail" ]; then
		echo "Test complete"
	else
		echo "Test failed"
		return 1
	fi
}
