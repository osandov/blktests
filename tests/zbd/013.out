Running zbd/013
Test complete
